src/canary.rs

```diff
@@ -233,56 +232,165 @@ macro_rules! assert_subroutine {
-        0xfb, 0xe7, // b.n      200ec <paint>
+        0xfb, 0xe7, // b.n      #0xfffffffa <paint>
         0x00, 0xbe, // bkpt     0x0000
         0x00, 0xbe, // bkpt     0x0000 (padding instruction)
     ];
 }
 
-/// Prepare target to execute subroutine.
+/// Measure-stack subroutine.
+///
+/// ### Corresponds to following rust code
 ///
-/// After calling this function, the program counter will be at the beginning of
-/// the subroutine.
+/// ```rust
+/// #[export_name = "measure"]
+/// unsafe fn measure(mut low_addr: u32, high_addr: u32, pattern: u32) -> u32 {
+///     let mut result = 0;
+///
+///     while low_addr < high_addr {
+///         if (low_addr as *const u32).read() != pattern {
+///             result = low_addr;
+///             break;
+///         } else {
+///             low_addr += 4;
+///         }
+///     }
+///
+///     result
+/// }
+/// ```
+///
+/// ### Generated assembly
+///
+/// The assembly is generated from aboves rust code, using the jorge-hack.
+///
+/// ```armasm
+/// 000200ec <measure>:
+///     200ec:    4288    cmp      r0, r1
+///     200ee:    d204    bcs.n    #0xc <measure+0xe>
+///     200f0:    6803    ldr      r3, [r0, #0]
+///     200f2:    4293    cmp      r3, r2
+///     200f4:    d102    bne.n    #8 <measure+0x10>
+///     200f6:    1d00    adds     r0, r0, #4
+///     200f8:    e7f8    b.n      #0xfffffff8 <measure>
+///
+/// 000200fa <measure+0xe>:
+///     200fa:    2000    movs     r0, #0
+///
+/// 000200fc <measure+0x10>:
+///     200fc:    be00    bkpt     0x0000
+/// //                    ^^^^ this was `bx lr`
+/// ```
+///
+/// ### Register-parameter-mapping
+///
+/// - r0: low_addr, and return value
+/// - r1: high_addr
+/// - r2: pattern
+mod measure_subroutine {
+    use super::*;
+
+    /// Search for lowest touched address in memory.
+    ///
+    /// The returned `Option<u32>` is `None`, if the memory is untouched. Otherwise it
+    /// gives the position of the lowest byte which isn't equal to the pattern anymore.
+    ///
+    /// ## Assumptions
+    /// - Expects the [`Core`] to be halted and will leave it halted when the function
+    /// returns.
+    /// - `low_addr` and `size` need to be 4-byte-aligned.
+    pub fn execute(
+        core: &mut Core,
+        low_addr: u32,
+        stack_size: u32,
+    ) -> Result<Option<u32>, probe_rs::Error> {
+        assert_subroutine!(low_addr, stack_size, self::SUBROUTINE.len() as u32);
+
+        // use probe to search through the memory the subroutine will be written to
+        match self::search_with_probe(core, low_addr)? {
+            addr @ Some(_) => return Ok(addr), // if we find a touched value, return early ...
+            None => {}                         // ... otherwise we continue
+        }
+
+        super::execute_subroutine(core, low_addr, stack_size, self::SUBROUTINE)?;
+        self::get_result(core)
+    }
+
+    fn search_with_probe(core: &mut Core, low_addr: u32) -> Result<Option<u32>, probe_rs::Error> {
```

Comment:
I'd like to see comments for these private functions, for the benefit of the next maintainer.

```
/// Searches though memory byte by byte using the SWD/JTAG probe.
///
/// We need this because we're about to overwrite some stack with a
/// canary-measuring program, but we need to check the memory we're about
/// to overwrite in case the stack usage got this far.
```

Reply:
I've documented all functions. What do you think?